src/Raven.Server/Documents/Handlers/QueriesHandler.cs

```diff
@@ -122,49 +124,45 @@ private void Explain(DocumentsOperationContext context, string indexName)
         public Task Patch()
         {
             DocumentsOperationContext context;
-            using (ContextPool.AllocateOperationContext(out context))
-            {
-                var reader = context.Read(RequestBodyStream(), "ScriptedPatchRequest");
-                var patch = PatchRequest.Parse(reader);
-
-                return ExecuteQueryOperation((runner, indexName, query, options, token) => runner.ExecutePatchQuery(indexName, query, options, patch, context, token), context);
-            }
+            ContextPool.AllocateOperationContext(out context); // we don't dispose this as operation is async
+            
+            var reader = context.Read(RequestBodyStream(), "ScriptedPatchRequest");
+            var patch = PatchRequest.Parse(reader);
+
+            ExecuteQueryOperation((runner, indexName, query, options, onProgress, token) => runner.ExecutePatchQuery(indexName, query, options, patch, context, onProgress, token), 
+                context, DatabaseOperations.PendingOperationType.UpdateByIndex);
+            return Task.CompletedTask;
         }
 
-        private async Task ExecuteQueryOperation(Func<QueryRunner, string, IndexQueryServerSide, QueryOperationOptions, OperationCancelToken, Task> operation, DocumentsOperationContext context)
+        private void ExecuteQueryOperation(Func<QueryRunner, string, IndexQueryServerSide, QueryOperationOptions, Action<IOperationProgress>, OperationCancelToken, Task<IOperationResult>> operation, DocumentsOperationContext context,
+            DatabaseOperations.PendingOperationType operationType)
         {
             var indexName = RouteMatch.Url.Substring(RouteMatch.MatchLength);
 
             var query = IndexQueryServerSide.Create(HttpContext, GetStart(), GetPageSize(int.MaxValue), context);
             var options = GetQueryOperationOptions();
             var token = CreateTimeLimitedOperationToken();
 
-            // TODO [ppekrol] 
-            // implement Tasks
-            // support RetrieveDetails
+            var queryRunner = new QueryRunner(Database, context);
 
-            using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
-            {
-                var queryRunner = new QueryRunner(Database, context);
-                await operation(queryRunner, indexName, query, options, token).ConfigureAwait(false);
+            long operationId = GetLongQueryString("operationId").Value;
 
-                writer.WriteStartObject();
+            var task = Database.DatabaseOperations.AddOperation(indexName, operationType, onProgress => 
+                    operation(queryRunner, indexName, query, options, onProgress, token), operationId, token);
 
-                writer.WritePropertyName("OperationId");
-                writer.WriteInteger(-1); // TODO [ppekrol]
-
-                writer.WriteEndObject();
-            }
+            task.ContinueWith(_ => context.Dispose());
```

Comment:
You shouldn't be disposing the context, you should be returning it to the pool.
In other words, you need to dispose the result of AllocateContext method.